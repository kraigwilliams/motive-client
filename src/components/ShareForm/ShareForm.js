import React, { Component } from 'react'
import { FormWrapper, FormTitle, FormLabel, Dropdown, DetailMessage} from '../Form/Form'
import {PageWrapper, colors} from '../constants'
import { FormButton, GoBack } from '../Button/Button'
import ActionsService from '../../services/actions-service'
import UserContext from '../../contexts/UserContext'

export default class ShareForm extends Component {
  // will have to use both contexts here to get the user id from user context and the thought id from the content context 
  static contextType = UserContext;

  constructor(props) {
    super(props)
    this.state = {
      share_id: null,
      item: null,
      connections: [],
      connectionSelected: {},
      shareSelected: null
    }
  }

  async componentDidMount() {
    if(this.props.match.params.thought_id) {
      this.setState({
        share_id : this.props.match.params.thought_id,
        item: 'thought'
      })
    } else {
      this.setState({
        share_id : this.props.match.params.topic_id,
        item: 'topic'
      })
    }
    
    const { user } = this.context;
    const userId = user.id;
    const connections = await ActionsService.getConnections(userId)
    this.setState({
      connections
    })
  }

  handleSelectedShareChange(ev) {
    ev.preventDefault()
    const shareSelected = ev.target.value;
    console.log(shareSelected)
    this.setState({
      shareSelected
    })
  }

  handleSubmitShare(ev) {
    ev.preventDefault()
    const { share_id, item } = this.state;
    const { connections, share_level } = ev.target;
    const connections_share = connections.value;
    const share_level_share = share_level.value;

    console.log(connections_share, 'connections', share_level_share, 'share level')
    if(item === 'thought') {
      ActionsService.shareThought(share_id, connections_share, share_level_share);
      this.props.history.goBack(`thoughts/${share_id}`)
    } else if (item === 'topic') {
      ActionsService.shareTopic(share_id, connections_share, share_level_share);
      this.props.history.goBack( `topics/${share_id}`)
    }
  }

  render() {
    const {connections, shareSelected} = this.state;
    const connectOptions = connections.map((friend, idx) => {
      return <option key={idx} value={friend.id}>
        {friend.first_name} 
        {' '}
        {friend.last_name} 
        {' '}
        {friend.username}
      </option>
    })
    return (
      <PageWrapper padding='40px 0' bgColor={colors.slategrey}>
        <FormWrapper padding='50px 30px' onSubmit={this.handleSubmitShare.bind(this)}>
          <GoBack
            type='reset'
            onClick={() => this.props.history.goBack()}
            margin='0px'
          />
          <FormTitle>
            Share
          </FormTitle>

          <FormLabel htmlFor='share-connections'>
            Share with:
          </FormLabel>
          <Dropdown
            id='share-connections'
            aria-label="You can select a connection to share this thought with"
            name='connections'
            defaultValue={0}
          >
            <option value={0}> -- Choose a connection -- </option>
            {/* options of all connections */}
            {connectOptions}
          </Dropdown>

          <FormLabel>
            Share as: 
          </FormLabel>
          <Dropdown
             id='share_level'
             aria-label="You can select what type of share you wish to do"
             name='share_level'
             defaultValue={0}
             onChange={this.handleSelectedShareChange.bind(this)}
          >
            {/* options to share as a collab or as reader */}
            <option value={0}>Choose how to share</option>
            <option value={2}>Collaborator</option>
            <option value={3}>Viewer</option>
          </Dropdown>

          <FormButton 
              type='submit'
            >
            Share
          </FormButton>

          {/* !!!! Conditionally render a message based on what the user has selected to describe the difference between a collaborator and a reader !!!! */}
          
          {
            Number(shareSelected) === 2 
              ? <DetailMessage> Sharing as a collaborator allows them to edit the thought you created </DetailMessage>
              : null
          }

          {
            Number(shareSelected) === 3 
              ?  <DetailMessage> Sharing as a reader only allows them to view and comment on the thought you created </DetailMessage>
              : null
          }
        </FormWrapper>
      </PageWrapper>

    )
  }
}