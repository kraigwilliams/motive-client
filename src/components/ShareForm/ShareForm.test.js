import React from "react";
import ReactDOM from "react-dom";
import ShareForm from "./ShareForm";
import renderer from "react-test-renderer";

describe(`<ShareForm />`, () => {
  //Smoke Testing
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<ShareForm />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  //Snapshot Testing
  it("renders UI as expected", () => {
    const tree = renderer.create(<ShareForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
