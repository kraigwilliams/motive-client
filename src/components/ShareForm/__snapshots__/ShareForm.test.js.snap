// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShareForm /> renders UI as expected 1`] = `
<div
  className="sc-AxjAm dtabnj"
>
  <form
    className="sc-AxiKw gwceYh"
    onSubmit={[Function]}
  >
    <button
      className="Button sc-fzoLsD irZtSf"
      color="#829191"
      margin="0px"
      onClick={[Function]}
      type="reset"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-arrow-left fa-w-14 "
        data-icon="arrow-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
    <h1
      className="sc-AxirZ gSjuDs"
    >
      Share
    </h1>
    <label
      className="Label sc-AxhUy iPWEFB"
      htmlFor="share-connections"
    >
      Share with:
    </label>
    <select
      aria-label="You can select a connection to share this thought with"
      className="sc-Axmtr ewTDfS"
      defaultValue={0}
      id="share-connections"
      name="connections"
    >
      <option
        value={0}
      >
         -- Choose a connection -- 
      </option>
    </select>
    <label
      className="Label sc-AxhUy iPWEFB"
    >
      Share as:
    </label>
    <select
      aria-label="You can select what type of share you wish to do"
      className="sc-Axmtr ewTDfS"
      defaultValue={0}
      id="share_level"
      name="share_level"
      onChange={[Function]}
    >
      <option
        value={0}
      >
        Choose how to share
      </option>
      <option
        value={2}
      >
        Collaborator
      </option>
      <option
        value={3}
      >
        Viewer
      </option>
    </select>
    <button
      className="Button sc-fzplWN yPljW"
      color="#EB9C82"
      type="submit"
    >
      Share
    </button>
  </form>
</div>
`;
